cpp/velox/shuffle/VeloxSortShuffleWriter.cc

```diff
@@ -0,0 +1,314 @@
+}
+} // namespace
+
+arrow::Result<std::shared_ptr<VeloxShuffleWriter>> VeloxSortShuffleWriter::create(
+    uint32_t numPartitions,
+    std::unique_ptr<PartitionWriter> partitionWriter,
+    ShuffleWriterOptions options,
+    std::shared_ptr<facebook::velox::memory::MemoryPool> veloxPool,
+    arrow::MemoryPool* arrowPool) {
+  std::shared_ptr<VeloxSortShuffleWriter> writer(new VeloxSortShuffleWriter(
+      numPartitions, std::move(partitionWriter), std::move(options), std::move(veloxPool), arrowPool));
+  RETURN_NOT_OK(writer->init());
+  return writer;
+}
+
+VeloxSortShuffleWriter::VeloxSortShuffleWriter(
+    uint32_t numPartitions,
+    std::unique_ptr<PartitionWriter> partitionWriter,
+    ShuffleWriterOptions options,
+    std::shared_ptr<facebook::velox::memory::MemoryPool> veloxPool,
+    arrow::MemoryPool* pool)
+    : VeloxShuffleWriter(numPartitions, std::move(partitionWriter), std::move(options), std::move(veloxPool), pool),
+      allocator_{std::make_unique<facebook::velox::HashStringAllocator>(veloxPool_.get())},
+      array_{SortArray{Allocator(allocator_.get())}} {}
+
+arrow::Status VeloxSortShuffleWriter::write(std::shared_ptr<ColumnarBatch> cb, int64_t memLimit) {
+  ARROW_ASSIGN_OR_RAISE(auto rv, getPeeledRowVector(cb));
+  initRowType(rv);
+  RETURN_NOT_OK(insert(rv, memLimit));
+  return arrow::Status::OK();
+}
+
+arrow::Status VeloxSortShuffleWriter::stop() {
+  ARROW_RETURN_IF(evictState_ == EvictState::kUnevictable, arrow::Status::Invalid("Unevictable state in stop."));
+
+  EvictGuard evictGuard{evictState_};
+
+  stopped_ = true;
+  RETURN_NOT_OK(evictAllPartitions());
+  RETURN_NOT_OK(partitionWriter_->stop(&metrics_));
+  return arrow::Status::OK();
+}
+
+arrow::Status VeloxSortShuffleWriter::reclaimFixedSize(int64_t size, int64_t* actual) {
+  if (evictState_ == EvictState::kUnevictable || offset_ == 0) {
+    *actual = 0;
+    return arrow::Status::OK();
+  }
+  EvictGuard evictGuard{evictState_};
+  auto beforeReclaim = veloxPool_->usedBytes();
+  RETURN_NOT_OK(evictAllPartitions());
+  *actual = beforeReclaim - veloxPool_->usedBytes();
+  return arrow::Status::OK();
+}
+
+arrow::Status VeloxSortShuffleWriter::init() {
+  ARROW_RETURN_IF(
+      options_.partitioning == Partitioning::kSingle,
+      arrow::Status::Invalid("VeloxSortShuffleWriter doesn't support single partition."));
+  array_.resize(initialSize_);
+  return arrow::Status::OK();
+}
+
+void VeloxSortShuffleWriter::initRowType(const facebook::velox::RowVectorPtr& rv) {
+  if (UNLIKELY(!rowType_)) {
+    rowType_ = facebook::velox::asRowType(rv->type());
+    fixedRowSize_ = facebook::velox::row::CompactRow::fixedRowSize(rowType_);
+  }
+}
+
+arrow::Result<facebook::velox::RowVectorPtr> VeloxSortShuffleWriter::getPeeledRowVector(
+    const std::shared_ptr<ColumnarBatch>& cb) {
+  if (options_.partitioning == Partitioning::kRange) {
+    auto compositeBatch = std::dynamic_pointer_cast<CompositeColumnarBatch>(cb);
```

Comment:
It seems that the casting highly depends on the preceding code calling `ColumnarBatch#compose`. Is it possible to directly call `VeloxColumnarBatch::from` which is more general, then operate on the output Velox columnar batch? Which shouldn't bring visible overhead based on my perspective.